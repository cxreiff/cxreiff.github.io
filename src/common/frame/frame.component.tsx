import {
  useRef,
  ReactNode,
  CSSProperties,
  useEffect,
  useLayoutEffect,
  useState,
} from "react";
import cn from "classnames";

import { IconButton } from "../iconButton/iconButton.component";

import styles from "./frame.module.scss";
import { MaximizeIcon, MinimizeIcon } from "lucide-preact";

type FrameProps = {
  aspect: number;
  loading?: boolean;
  canFullscreen?: boolean;
  externalLink?: string;
  className?: string;
  style?: CSSProperties;
  children: ReactNode;
};

export const Frame = ({
  aspect,
  loading = false,
  canFullscreen = false,
  className,
  style,
  children,
}: FrameProps) => {
  const ref = useRef<HTMLDivElement>(null);
  const [fullscreen, setFullscreen] = useState(false);

  useEffect(() => {
    if (fullscreen && document.fullscreenElement == null) {
      ref.current?.requestFullscreen();
    } else if (!fullscreen && document.fullscreenElement != null) {
      document.exitFullscreen();
    }
  }, [fullscreen]);

  useLayoutEffect(() => {
    document.onfullscreenchange = () =>
      setFullscreen(document.fullscreenElement != null);
    return () => {
      document.onfullscreenchange = null;
    };
  });

  return (
    <div
      className={cn(styles.frame, className, { [styles.hidden]: loading })}
      style={style}
      ref={ref}
    >
      {children}
      <svg viewBox={`0 0 ${aspect} 1`} />
      {canFullscreen && document.fullscreenEnabled && (
        <IconButton
          className={styles.fullscreen_button}
          onClick={() => setFullscreen(!fullscreen)}
        >
          {fullscreen ? <MinimizeIcon /> : <MaximizeIcon />}
        </IconButton>
      )}
    </div>
  );
};
