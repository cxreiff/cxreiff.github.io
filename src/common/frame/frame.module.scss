@use '../../shared.scss' as *;

.frame {
  position: relative;
  width: 100%;
  height: 100%;
  overflow: hidden;

  &>svg {
    position: relative;
  }

  &>.content {
    max-width: 100%;
    max-height: 100%;
    position: absolute;
    z-index: 1;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    margin: auto;
  }

  &.hidden>.content {
    display: none;
  }

  &:not(.hidden)>.content {
    animation: fadeIn ease-in 0.3s;
  }
}

@keyframes fadeIn {
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}
