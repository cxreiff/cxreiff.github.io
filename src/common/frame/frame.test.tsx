import { test, expect } from "vitest";
import { render } from "@testing-library/preact";

import { Frame } from "./frame.component";
import styles from "./frame.module.scss";

test("should render", () => {
  const { container, getByAltText } = render(
    <Frame
      aspect={16 / 9}
      loading={false}
      className={"test"}
      style={{ background: "red" }}
    >
      <img src={"test"} alt={"test"} />
    </Frame>
  );

  expect(getByAltText("test")).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});

test("should constrain size to provided aspect ratio", () => {
  const { container, rerender } = render(
    <Frame aspect={9 / 16}>
      <img src={"test"} alt={"test"} />
    </Frame>
  );

  expect(container.firstElementChild?.lastElementChild).toBeInTheDocument();
  expect(
    container.firstElementChild!.lastElementChild!.getAttribute("viewBox")
  ).toEqual(`0 0 ${9 / 16} 1`);

  rerender(
    <Frame aspect={16 / 9}>
      <img src={"test"} alt={"test"} />
    </Frame>
  );

  expect(container.firstElementChild?.lastElementChild).toBeInTheDocument();
  expect(
    container.firstElementChild!.lastElementChild!.getAttribute("viewBox")
  ).toEqual(`0 0 ${16 / 9} 1`);
});

test("should hide when loading prop is true", () => {
  const { container } = render(
    <Frame aspect={9 / 16} loading={true}>
      <img src={"test"} alt={"test"} />
    </Frame>
  );

  expect(container.firstElementChild).toBeInTheDocument();
  expect(container.firstElementChild!).toHaveClass(styles.hidden);
  expect(container).toMatchSnapshot();
});
