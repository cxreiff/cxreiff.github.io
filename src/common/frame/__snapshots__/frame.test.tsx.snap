// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should hide when loading prop is true 1`] = `
<div>
  <div
    class="_frame_331539 _hidden_331539"
  >
    <img
      alt="test"
      src="test"
    />
    <svg
      viewBox="0 0 0.5625 1"
    />
  </div>
</div>
`;

exports[`should render 1`] = `
<div>
  <div
    class="_frame_331539 test"
    style="background: red;"
  >
    <img
      alt="test"
      src="test"
    />
    <svg
      viewBox="0 0 1.7777777777777777 1"
    />
  </div>
</div>
`;
